### PR TITLE
41 - scaled to percents

### PR DESCRIPTION
**Pull request by kolesroma**
todo write to database

### CODE REVIEW

File: src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java [hunk 1]
```diff
@@ -14,9 +14,9 @@
 @RequiredArgsConstructor
 public class QuizResultServiceImpl implements QuizResultService {
     @Override
-    public Map<Long, Integer> calculateResult(List<EmotionDto> emotions) {
+    public Map<Long, Double> calculateResult(List<EmotionDto> emotions) {
         return emotions.stream()
                 .collect(Collectors.groupingBy(EmotionDto::getEmotionId, LinkedHashMap::new,
-                        Collectors.summingInt(EmotionDto::getValue)));
+                        Collectors.averagingInt(emotionDto -> emotionDto.getValue() * 20)));
```
Comment (maximelyashevich) [inline src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java:20]: What is a magic number 20 - constant?

**Review by maximelyashevich — COMMENTED** [review]
Hi @kolesroma,
Please look through my comment.

Regards,
Maks

File: src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java (hunk 1 above)
Comment (kolesroma) [reply src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java:20]: Hi @maximelyashevich . We multiply interval [0;5] by koeficient 20 and get a new one [0;100], which represents percents. 0-5 is a value of user emotion we get from FE

Comment (maximelyashevich) [reply src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java:20]: Thanks for the provided details - let's add a constant in this class with this value.

File: src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java [hunk 2]
```diff
@@ -13,10 +20,26 @@
 @Service
 @RequiredArgsConstructor
 public class QuizResultServiceImpl implements QuizResultService {
+    private final QuizResultRepository quizResultRepository;
+    private final EmotionService emotionService;
+    private final QuizService quizService;
+    private static final int percentCoefficient = 20;
```
Comment (maximelyashevich) [inline src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java:26]:
```suggestion
    private static final int PERCENT_COEFFICIENT = 20;
```

File: src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java [hunk 3]
```diff
@@ -13,10 +20,26 @@
...
+                            .user(null)
+                            .emotion(new Emotion(em.getEmotionId(), null))
+                            .score(em.getValue())
+                            .createdAt(LocalDateTime.now())
+                            .build();
+                    quizResultRepository.saveAndFlush(quizResult);
+                })
                 .collect(Collectors.groupingBy(EmotionDto::getEmotionId, LinkedHashMap::new,
-                        Collectors.summingInt(EmotionDto::getValue)));
+                        Collectors.averagingInt(emotionDto -> emotionDto.getValue() * percentCoefficient)));
```
Comment (maximelyashevich) [inline src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java:43]:
```suggestion
                        Collectors.averagingInt(emotionDto -> emotionDto.getValue() * PERCENT_COEFFICIENT)));
```

**Review by maximelyashevich — COMMENTED** [review]
Please review.

File: src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java [hunk 4]
```diff
@@ -13,10 +18,27 @@
...
+    private static final int PERCENT_COEFFICIENT = 20;
+
+    @Transactional
+    @Override
+    public void create(List<EmotionDto> emotions) {
+        Long resultId = quizResultRepository.getNextRowId();
+        emotions.forEach(dto -> quizResultRepository.save(QuizResult.builder()
+                .id(resultId)
+                .quiz(null)
+                .user(null)
```
Comment (maximelyashevich) [inline src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java:31]: I think we can ignore that fields.

File: src/main/java/com/epam/emotionalhelp/repository/QuizResultRepository.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+import java.util.Collection;
+
+public interface QuizResultRepository extends JpaRepository<QuizResult, Long> {
+    Collection<QuizResult> findAllById(Long resultId);
+
+    @Query(value = "SELECT MAX(id) + 1 FROM QuizResult")
+    Long getNextRowId();
```
Comment (maximelyashevich) [inline src/main/java/com/epam/emotionalhelp/repository/QuizResultRepository.java:13]: We have agreed with @GeloPakDev to create a new `quiz_attempt` table that will store actual quiz attempt - please consider it in your implementation if needed.

See [Database model](https://github.com/emotional-help-competition/backend/wiki/Database-model) for more details.

File: src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java [hunk 6]
```diff
@@ -13,10 +18,27 @@
 @Service
 @RequiredArgsConstructor
 public class QuizResultServiceImpl implements QuizResultService {
+    private final QuizResultRepository quizResultRepository;
+    private static final int PERCENT_COEFFICIENT = 20;
```
Comment (maximelyashevich) [inline src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java:22]:
```suggestion
    private static final int PERCENT_COEFFICIENT = 20;
    
    private final QuizResultRepository quizResultRepository;
```
Generally, we can declare constants - first and final fields - next.

File: src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java [hunk 7]
```diff
@@ -13,10 +18,27 @@
...
+    @Override
+    public void create(List<EmotionDto> emotions) {
+        Long resultId = quizResultRepository.getNextRowId();
+        emotions.forEach(dto -> quizResultRepository.save(QuizResult.builder()
+                .id(resultId)
+                .quiz(null)
+                .user(null)
+                .emotion(new Emotion(dto.getEmotionId(), null))
+                .score(dto.getValue())
+                .createdAt(LocalDateTime.now())
```
Comment (maximelyashevich) [inline src/main/java/com/epam/emotionalhelp/service/impl/QuizResultServiceImpl.java:34]: Just a proposal:
Can we consider JPA entity listener to set created date?

**Review by maximelyashevich — COMMENTED** [review]
Hi @kolesroma,
Please consider what can be improved in your code changes.

Regards,
Maks

**Comment by kolesroma** [comment]
changed the database model - no need for my pr

**Pull request closed without merging**